### ISSUE TITLE
Islands have no memory

### ISSUE THREAD

**Issue by jakobj** [esa/pagmo2#181]
When calling `evolve()` repeatedly on an `island` instance, I expected the optimization to continue if the corresponding flag in the algorithm was set. See the following example:
```python
import pygmo as pg

if __name__ == '__main__':
    algo = pg.algorithm(pg.xnes(gen=50, memory=True, seed=123))
    algo.set_verbosity(10)
    prob = pg.problem(pg.rosenbrock(10))
    pop = pg.population(prob, size=20, seed=123)
    isl = pg.island(algo=algo, pop=pop, udi=pg.mp_island())

    isl.evolve()
    isl.wait()
    isl.evolve()
    isl.wait()
```
The optimization, however, starts from new, random initial conditions. Is this is the indented behavior or a bug? If this is intended, is there any way to make the islands have memory across multiple calls to evolve?

**Comment by darioizzo** [comment]
No, it is not the intended behaviour. 
The flag memory refers to the algorithms internals anyway, not to the population. The population should not be reset between successive calls, and it is not on my system.

 Can you produce the output of your code that contradicts the intended behaviour?

**Comment by jakobj** [comment]
Yes, you are right, actually I had two expectations:
1) the `memory=True` flag makes sure the algorithm internals, e.g., mean/std of the search distribution in xnes are not reset
2) the population is not reset
However, running the above code produces the following output on my machine:
```bash
xNES 4 PaGMO:
eta_mu: 1 - eta_sigma: 0.100609 - eta_b: 0.100609 - sigma0: 0.5
utilities: [0.220199, 0.142094, 0.0964057, 0.0639891, 0.0388449, ... ]

   Gen:        Fevals:          Best:            dx:            df:         sigma:
      1             20    3.27173e+06        28.6374    6.06109e+07            0.5
     11            220         547317        16.4028    9.28074e+06       0.407837
     21            420         230381        12.0726    6.06792e+06        0.34032
     31            620         146936        11.0385    2.15225e+06       0.284385
     41            820        34496.2        11.7758    2.91558e+06       0.231816
Exit condition -- generations = 50
xNES 4 PaGMO:
eta_mu: 1 - eta_sigma: 0.100609 - eta_b: 0.100609 - sigma0: 0.5
utilities: [0.220199, 0.142094, 0.0964057, 0.0639891, 0.0388449, ... ]

   Gen:        Fevals:          Best:            dx:            df:         sigma:
      1             20         935053        28.6374     5.6233e+07            0.5
     11            220         484706        16.7617    9.63572e+06       0.407733
     21            420         271813        12.1776    6.21324e+06       0.341389
     31            620         132106        11.0841    2.06378e+06       0.283816
     41            820        39422.8        11.8588    2.74919e+06       0.231582
Exit condition -- generations = 50
```
From this I would conclude that the optimization is basically restarted instead of continuing the previous `evolve`. Is this different on your system?

**Comment by jakobj** [comment]
Any idea what I am doing wrong?

**Comment by darioizzo** [comment]
So, we investigated and @bluescarni correctly pointed out that in the ```pg.mp_island()``` the algorithm is not returned, only the population. Hence each evolve is not remembering the algo state and that is why xnes starts from scratch (note that the population is changed and remembered, only the algo state is not)

What are you doing wrong? Nothing. We have to change the behaviour of the isands. In the meanwhile, possible workarounds

1. Do not use the mp islands. ```pop = algo.evolve(pop)``` will work. Also using the thread island will work but its not possible on python written algo or probs
2. Use other algorithms, the ones without memory. 

Let us know.

**Comment by jakobj** [comment]
OK, thanks for clarifying this. Unfortunately I currently need both multiprocessing and xnes, so I will for the time being rely on my custom implementation. Do you have any estimate when this might be fixed?

**Comment by darioizzo** [comment]
Another possible workaround is to let xnes work for more generation and call one evolve only (with 100 gen).

The fix will probably be late after summer :( sorry.

**Comment by jakobj** [comment]
Yes, that would be one option. However I would like to run evaluation runs multiple times during the evolution, would like to checkpoint the states etc., which requires me to stop and continue the evolution many times.

No problem, thanks for clarifying and I'll make sure to check back later this summer.

**Comment by bluescarni** [comment]
I have a PR up with a tentative fix at #186.

**Comment by darioizzo** [comment]
I can confirm the issue is solved, I now get this output from the user example, which shows how the memory is kept after the first call to evolve

```
xNES 4 PaGMO: 
eta_mu: 1 - eta_sigma: 0.100609 - eta_b: 0.100609 - sigma0: 0.5
utilities: [0.220199, 0.142094, 0.0964057, 0.0639891, 0.0388449, ... ]

   Gen:        Fevals:          Best:            dx:            df:         sigma:
      1             20    3.27173e+06        28.6374    6.06109e+07            0.5
     11            220         547317        16.4028    9.28074e+06       0.407837
     21            420         230381        12.0726    6.06792e+06        0.34032
     31            620         146936        11.0385    2.15225e+06       0.284385
     41            820        34496.2        11.7758    2.91558e+06       0.231816
Exit condition -- generations = 50
xNES 4 PaGMO: 
eta_mu: 1 - eta_sigma: 0.100609 - eta_b: 0.100609 - sigma0: 0.188805
utilities: [0.220199, 0.142094, 0.0964057, 0.0639891, 0.0388449, ... ]

   Gen:        Fevals:          Best:            dx:            df:         sigma:
      1             20        27338.7        7.81385         269083       0.188805
     11            220        17734.2        5.98226         352854       0.159852
     21            420        8205.17        8.76039         129895       0.133659
     31            620         3440.5        6.34298        90996.4       0.114103
     41            820        1776.55        5.17037          21135      0.0935456
Exit condition -- generations = 50

```

**Comment by bluescarni** [comment]
I have merged the PR to master, we'll probably have a 2.8 release soonish.

**Comment by bluescarni** [comment]
I am closing the issue as it looks like the problem has been solved by #186.
